pkg/auth/providers/saml/saml_client.go

```diff
@@ -0,0 +1,277 @@
+package saml
+
+import (
+	"bufio"
+	"crypto/rsa"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"encoding/pem"
+	"encoding/xml"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/crewjam/saml"
+	"github.com/crewjam/saml/samlsp"
+	"github.com/gorilla/mux"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	log "github.com/sirupsen/logrus"
+)
+
+//SPClient implements a client for the saml library
+type SPClient struct {
+	config *v3.SamlConfig
+	SamlSP *samlsp.Middleware
+}
+
+type IDPMetadata struct {
+	XMLName           xml.Name                `xml:"urn:oasis:names:tc:SAML:2.0:metadata EntityDescriptor"`
+	ValidUntil        time.Time               `xml:"validUntil,attr"`
+	EntityID          string                  `xml:"entityID,attr"`
+	IDPSSODescriptors []saml.IDPSSODescriptor `xml:"IDPSSODescriptor"`
+	SPSSODescriptors  []saml.SPSSODescriptor  `xml:"SPSSODescriptor"`
+}
+
+const (
+	redirectBackBase = "redirectBackBase"
+	redirectBackPath = "redirectBackPath"
+)
+
+var Root *mux.Router
+
+func (s *Provider) InitializeSPClient(configToSet *v3.SamlConfig, name string) error {
```

Comment:
As mentioned elsewhere, this should not be attached to any struct.
Also, can you change its name? it is not initializing a client.
Can you also move it to a different file?